test/integration/admin/api/registry/policies_controller_test.rb

```diff
@@ -155,6 +155,112 @@ def setup
   end
 
+  class PolicyInUseTest < ActionDispatch::IntegrationTest
+    def setup
+      @provider = FactoryBot.create(:provider_account)
+      policy_schema = JSON.parse(file_fixture('policies/apicast-policy.json').read)
+      @policy = FactoryBot.create(:policy, account: @provider,
+                                           name: policy_schema['name'],
+                                           version: policy_schema['version'],
+                                           schema: policy_schema)
+      host! @provider.admin_domain
+      @access_token = FactoryBot.create(:access_token, owner: @provider.admin_users.first!, scopes: %w[policy_registry], permission: 'rw')
+      ::Account.any_instance.stubs(:provider_can_use?).returns(true)
+    end
+
+    attr_reader :provider, :policy, :access_token
+    delegate :schema, to: :policy, prefix: true
+    delegate :proxy, to: 'provider.default_service'
+
+    test 'can only update policy if not in use by any proxy' do
+      try_update_policy version: '1.1', schema: policy_schema.merge('version': '1.1').to_json
+      assert_response :success
+
+      policy.reload
+
+      add_policy_config_to(proxy)
+
+      try_update_policy version: '1.2', schema: policy_schema.merge('version': '1.2').to_json
+      assert_response :unprocessable_entity
+      assert_equal ['Cannot modify policy currently in use.'], JSON.parse(response.body).dig('errors', 'base')
+    end
+
+    test 'can only delete policy if not in use by any proxy' do
+      add_policy_config_to(proxy)
+
+      try_delete_policy
+      assert_response :forbidden
+      assert_equal ['Cannot modify policy currently in use.'], JSON.parse(response.body).dig('errors', 'base')
+
+      clear_policy_config_from(proxy)
+
+      try_delete_policy
+      assert_response :success
+    end
+
+    test 'one proxy using the policy is enough to forbid modifications' do
+      add_policy_config_to(proxy)
+
+      other_service = FactoryBot.create(:service, account: provider)
+      other_proxy = other_service.proxy
+      add_policy_config_to(other_proxy)
+
+      try_delete_policy
+      assert_response :forbidden
+      assert_equal ['Cannot modify policy currently in use.'], JSON.parse(response.body).dig('errors', 'base')
+
+      clear_policy_config_from(proxy)
+
+      try_delete_policy
+      assert_response :forbidden
+      assert_equal ['Cannot modify policy currently in use.'], JSON.parse(response.body).dig('errors', 'base')
+
+      clear_policy_config_from(other_proxy)
+
+      try_delete_policy
+      assert_response :success
+    end
+
+    test 'policies are not mixed up with other providers' do
+      add_policy_config_to(proxy)
+
+      other_provider = FactoryBot.create(:provider_account)
+      FactoryBot.create(:policy, account: other_provider, name: policy_schema['name'], version: policy_schema['version'], schema: policy_schema)
+      other_proxy = other_provider.default_service.proxy
+      add_policy_config_to(other_proxy)
```

Comment:
There's a bug here. I was suppose to add the policy created in line 228, not the one created in the setup. It won't change the result of the test, but I will fix it anyway.